x-pack/filebeat/input/awss3/collector.go

```diff
@@ -93,7 +130,260 @@ type s3Context struct {
+			if err != nil {
+				c.logger.Error(fmt.Errorf("handleS3Objects failed: %w", err))
+				break processingLoop
+			}
+		case err := <-errChan:
+			if err != nil {
+				c.logger.Error("S3 get objects failed: ", err)
+				break processingLoop
+			}
+		}
+	}
+
+	return
+}
+
+func (c *s3BucketCollector) purge() {
+	for _, listingID := range c.states.GetListingIDs() {
+		// workersListingMap map is shared with the storedop
+		// we lock on listing page and unlock when full listing page
+		// objects are stored (ie: acked)
+		// we lock here in order to process the purge only after
+		// full listing page acked
+		lock, _ := c.workersListingMap.Load(listingID)
+		lock.(*sync.Mutex).Lock()
+
+		keys := map[string]struct{}{}
+		latestStoredTimeByBucket := make(map[string]time.Time, 0)
+
+		for _, state := range c.states.GetStatesByListingID(listingID) {
+			// it is not stored, keep
+			if !state.Stored {
+				continue
+			}
+
+			var latestStoredTime time.Time
+			keys[state.Id] = struct{}{}
+			latestStoredTime, ok := latestStoredTimeByBucket[state.Bucket]
+			if !ok {
+				var commitWriteState commitWriteState
+				err := c.store.Get(awsS3WriteCommitPrefix+state.Bucket, &commitWriteState)
+				if err == nil {
+					// we have no entry in the map and we have no entry in the store
+					// set zero time
+					latestStoredTime = time.Time{}
+				}
+
+				latestStoredTime = commitWriteState.Time
+
+			}
+
+			if state.LastModified.After(latestStoredTime) {
+				latestStoredTimeByBucket[state.Bucket] = state.LastModified
+			}
+
+		}
+
+		for key := range keys {
+			c.states.Delete(key)
+		}
+
+		c.states.writeStates(c.store)
+
+		for bucket, latestStoredTime := range latestStoredTimeByBucket {
+			if err := c.store.Set(awsS3WriteCommitPrefix+bucket, commitWriteState{latestStoredTime}); err != nil {
+				c.logger.Errorf("Failed to write commit time to the registry: %+v", err)
+			}
+		}
+
+		// workersListingMap map is shared with the storedop
+		// purge is done, we can unlock and clean
+		lock.(*sync.Mutex).Unlock()
+		c.workersListingMap.Delete(listingID)
+		c.states.DeleteListing(listingID)
+	}
+
+	return
+}
+
+func (c *s3BucketCollector) run() {
+	defer c.logger.Info("s3 input workers have stopped.")
+	c.logger.Info("s3 input workers have started.")
+	for c.cancellation.Err() == nil {
+		// "sleep" for S3BucketPollInterval duration to reduce list operations
+		<-time.After(c.config.S3BucketPollInterval)
+		wg := new(sync.WaitGroup)
+
+		s3InfoChan := make(chan *s3Info)
+		errChan := make(chan error)
+
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, s3InfoChan chan<- *s3Info, errChan chan<- error) {
+			defer wg.Done()
+			c.getS3Objects(s3InfoChan, errChan)
+		}(wg, s3InfoChan, errChan)
+
+		for i := 0; i < c.config.S3BucketNumberOfWorkers; i++ {
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, s3InfoChan <-chan *s3Info, errChan <-chan error) {
+				defer wg.Done()
+				c.process(s3InfoChan, errChan)
+				c.purge()
```

Comment:
`purge` is concurrency safe, the lock is acquired in `getS3Objects` (https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R162) and released either in the same method (if no objects were collected: https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-0717dc8e6158cdc32ee79b54fe8fe67ea2ebad820d5935016e4f9a18d5cdd153R228) or when all the objects in the listing are stored (https://github.com/elastic/beats/pull/26934/files/05cd09612393dd9213aa2ac794595957f89bcd88#diff-34a1e64be4068f5099b1af6dc062ac9afd7f3f229041e25caeba0efa7f475e20R52)

Probably `purge` could be called after all the worker goroutine exits, since we will have the state in memory and the same object from different listing won't be processed twice anyway, but if filebeat dies or is restarted when some of the goroutine exited and some not yet we will not have the states synched to disk so the objects would be processed again when filebeat is restarted.

I was thinking about how much performance impact has synching the store to disk so often. Once moved the logic on your PR I will write a benchmark and see.